### PR TITLE
Make it easier to make optimizers: Move join input swapping and related methods into PhysicalOperators

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes #.

## Rationale for this change

I want to make it as easy as possible for people to implement (join order) optimizers on top of DataFusion (for example, https://github.com/cmu-db/optd, though I fully admit I have failed so far to get them to engage with us)

However, some of the key building blocks of such optimizations like swapping join inputs, are crate internal APIs of the very basic join ordering optimizer included with DataFusion. Thus it will be hard for others to find and use this cdoe

Some additional evidence there is demand for these functions is 
- this PR (that makes them more visible for use in forks) https://github.com/apache/datafusion/pull/13893 from @berkaysynnada 
- `swap_hash_join` is used in comet: https://github.com/apache/datafusion/pull/13899

Making the functions public and on the `Exec`s they are used for will make it even easier to find and use

This also makes the datafusion core crate a tiny bit smaller which should help compile times

## What changes are included in this PR?

There are no functional changes, this PR just moves code around

1. Create HashJoinExec::swap_inputs, `NestedLoopsJoinExec::swap_inputs`, and `CrossJoinExec::swap_inputs`
2. Refactor JoinSelection to use those functions
2. Extract JoinFilter into its own module, and move `swap` and `can_swap` there
4. Deprecate existing functions

### CODE REVIEW

File: datafusion/common/src/join_type.rs [hunk 1]
```diff
@@ -73,6 +73,41 @@ impl JoinType {
     pub fn is_outer(self) -> bool {
         self == JoinType::Left || self == JoinType::Right || self == JoinType::Full
     }
+
+    /// returns the new join type we get after swapping the given
+    /// join's inputs.
+    ///
+    /// Panics if self.supports_swap() returns false
+    pub fn swap(&self) -> JoinType {
```
Comment (alamb) [inline datafusion/common/src/join_type.rs:80]: These functions are moved from crate public functions in join_selection

File: datafusion/core/src/physical_optimizer/join_selection.rs [hunk 2]
```diff
@@ -108,197 +105,49 @@ fn supports_collect_by_thresholds(
 }
 
 /// Predicate that checks whether the given join type supports input swapping.
+#[deprecated(since = "45.0.0", note = "use JoinType::supports_swap instead")]
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/join_selection.rs:108]: This code is moved elsewhere, I chose to mark it as deprecated to help my migration as well as forks that might use these functions (e.g. Synnadas)

File: datafusion/core/src/physical_optimizer/join_selection.rs [hunk 3]
```diff
@@ -108,197 +105,49 @@ fn supports_collect_by_thresholds(
...
-        Ok(Arc::new(new_join))
-    } else {
-        // TODO avoid adding ProjectionExec again and again, only adding Final Projection
-        let proj = ProjectionExec::try_new(
-            swap_reverting_projection(&left.schema(), &right.schema()),
-            Arc::new(new_join),
-        )?;
-        Ok(Arc::new(proj))
-    }
+    hash_join.swap_inputs(partition_mode)
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/join_selection.rs:130]: This is the key point of this PR -- to move this (and related) function into the HashJoinExec itself

File: datafusion/core/src/physical_optimizer/join_selection.rs [hunk 4]
```diff
@@ -383,10 +232,10 @@ pub(crate) fn try_collect_left(
 
     match (left_can_collect, right_can_collect) {
         (true, true) => {
-            if supports_swap(*hash_join.join_type())
+            if hash_join.join_type().supports_swap()
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/join_selection.rs:235]: I think this code now reads more easily

File: datafusion/physical-plan/src/joins/cross_join.rs [hunk 5]
```diff
@@ -168,6 +169,19 @@ impl CrossJoinExec {
             boundedness_from_children([left, right]),
         )
     }
+
+    /// Returns a new `ExecutionPlan` that computes the same join as this one,
+    /// with the left and right inputs swapped using the  specified
+    /// `partition_mode`.
+    pub fn swap_inputs(&self) -> Result<Arc<dyn ExecutionPlan>> {
```
Comment (alamb) [inline datafusion/physical-plan/src/joins/cross_join.rs:175]: this was just refactored out of the swap rule

File: datafusion/physical-plan/src/joins/join_filter.rs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+use crate::joins::utils::ColumnIndex;
+use arrow_schema::Schema;
+use datafusion_common::JoinSide;
+use datafusion_physical_expr_common::physical_expr::PhysicalExpr;
+use std::sync::Arc;
+
+/// Filter applied before join output. Fields are crate-public to allow
+/// downstream implementations to experiment with custom joins.
+#[derive(Debug, Clone)]
+pub struct JoinFilter {
```
Comment (alamb) [inline datafusion/physical-plan/src/joins/join_filter.rs:27]: join/utils.rs is already quite large, so I moved this code into its own module

File: datafusion/physical-plan/src/joins/utils.rs [hunk 7]
```diff
@@ -1788,6 +1730,32 @@ impl BatchTransformer for BatchSplitter {
     }
 }
 
+/// When the order of the join is changed, the output order of columns must
+/// remain the same.
+///
+/// Returns the expressions that will allow to swap back the values from the
+/// original left as the first columns and those on the right next.
+pub(crate) fn swap_reverting_projection(
```
Comment (alamb) [inline datafusion/physical-plan/src/joins/utils.rs]: moved from optimizer utils

**Comment by alamb** [comment]
@berkaysynnada and @ozankabak  I wonder what you think about this PR (make the API for swapping join order easier / more public) given your comments on https://github.com/apache/datafusion/pull/13893#pullrequestreview-2521918076 that suggest these are widely used functions

**Comment by ozankabak** [comment]
Will review today

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 8]
```diff
@@ -566,8 +566,87 @@ impl HashJoinExec {
...
+    /// with the left and right inputs swapped using the  specified
+    /// `partition_mode`.
+    ///
+    /// # Notes:
+    ///
+    /// This function is public so other downstream projects can use it to
+    /// construct `HashJoinExec` with right side as the build side.
+    pub fn swap_inputs(
+        &self,
+        partition_mode: PartitionMode,
```
Comment (berkaysynnada) [inline datafusion/physical-plan/src/joins/hash_join.rs:580]: Isn't this a parameter which is present in the HashJoinExec itself? Why are we separately sending it 🤔

**Review by berkaysynnada — APPROVED** [review]
LGTM. I have just one question, but it is not much related with the scope of this PR. If that is a real defect, we can submit an issue. I will merge this soon if no further commits or comments.

File: datafusion/physical-plan/src/joins/hash_join.rs (hunk 8 above)
Comment (alamb) [reply datafusion/physical-plan/src/joins/hash_join.rs:580]: I am not 100% sure

Some uses clearly are unnecessary where the type of the join is simply passed in: https://github.com/apache/datafusion/blob/fce3fb3033ae50e8590e075ffb11422112128233/datafusion/core/src/physical_optimizer/join_selection.rs#L569

Other places like this are less clear that the output type is the same as the input type:
https://github.com/apache/datafusion/blob/fce3fb3033ae50e8590e075ffb11422112128233/datafusion/core/src/physical_optimizer/join_selection.rs#L285

I'll try and make a PR and see what happens if we don't send it

Comment (alamb) [reply datafusion/physical-plan/src/joins/hash_join.rs:580]: - I made https://github.com/apache/datafusion/pull/13940

the sqllogictests seem to have passed. I'll check the rest of CI in a bit

**Pull request merged**